lib/Conversion/LinalgToArith/LinalgToArith.cpp

```diff
@@ -0,0 +1,385 @@
+  } else {  // right operand is secret
+    return (i % dim0) * dim0 + ((i + j) % dim1);
+  }
+}
+
+struct ConvertLinalgMatmul : public OpRewritePattern<mlir::linalg::MatmulOp> {
+ private:
+  DataFlowSolver *solver;
+
+ public:
+  ConvertLinalgMatmul(DataFlowSolver *solver, mlir::MLIRContext *context)
+      : OpRewritePattern<mlir::linalg::MatmulOp>(context), solver(solver) {}
+
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::linalg::MatmulOp op,
+                                PatternRewriter &rewriter) const override {
+    // Determine if the left or right operand is secret to determine which
+    // matrix to diagonalize, or if both are secret or both are public, then
+    // return failure.
+    auto *leftOperandLookup =
+        solver->lookupState<SecretnessLattice>(op.getInputs()[0]);
+    Secretness leftOperandSecretness =
+        leftOperandLookup ? leftOperandLookup->getValue() : Secretness();
+    bool leftOperandIsSecret = (leftOperandSecretness.isInitialized() &&
+                                leftOperandSecretness.getSecretness());
+
+    auto *rightOperandLookup =
+        solver->lookupState<SecretnessLattice>(op.getInputs()[1]);
+    Secretness rightOperandSecretness =
+        rightOperandLookup ? rightOperandLookup->getValue() : Secretness();
+    bool rightOperandIsSecret = (rightOperandSecretness.isInitialized() &&
+                                 rightOperandSecretness.getSecretness());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Left operand is secret: " << leftOperandIsSecret << "\n"
+                   << "Right operand is secret: " << rightOperandIsSecret
+                   << "\n";
+    });
+
+    // Error out if both are secret or both are public
+    if ((leftOperandIsSecret && rightOperandIsSecret) ||
+        (!leftOperandIsSecret && !rightOperandIsSecret)) {
+      return failure();
+    }
+    auto inputs = op.getInputs();
+    auto outputs = op.getOutputs();
+
+    Value secretValues;
+    Value publicMatrix;
+    if (leftOperandIsSecret) {
+      secretValues = inputs[0];
+      publicMatrix = inputs[1];
+    } else {  // the right operand is secret
+      secretValues = inputs[1];
+      publicMatrix = inputs[0];
+    }
+    auto matrixTensorType = cast<RankedTensorType>(publicMatrix.getType());
+    auto bias = outputs[0];
+
+    auto dimensions = matrixTensorType.getShape();
+    int64_t dim0 = dimensions[0];  // This is the number of rows in the matrix.
+    int64_t dim1 = dimensions[1];  // This is the number of columns in the
+    // matrix.
+
+    SmallVector<int64_t> transposedDimensions({dim1, dim0});
+
+    // If one of these dimensions is not a power of two, then we can't do the
+    // Halevi-Shoup or Squat Packing Matrix Multiplication conversion.
+    if (!isPowerOfTwo(dim0) || !isPowerOfTwo(dim1)) {
+      return failure();
+    }
+
+    // If the matrix is not a square matrix, then we are doing squat packing.
```

Comment:
Error out here if since this only handles square matrices for now? (You can keep the comment to add squat support)